sw-emulator/lib/cpu/src/cpu.rs

```diff
@@ -262,7 +401,24 @@ impl<TBus: Bus> Cpu<TBus> {
     ///
     /// * `RvException` - Exception with cause `RvExceptionCause::IllegalRegister`
     pub fn write_xreg(&mut self, reg: XReg, val: RvData) -> Result<(), RvException> {
-        self.xregs.write(reg, val)
+        // XReg::X2 is the sp register.
+        if reg == XReg::X2 {
```

Comment:
Theoretically, we can set the `sp` to another value as a temporary, or when initializing (we set it to 0 at the beginning of the ROM), and this isn't necessarily a stack overflow, so we might have to be a little careful here.

Reply:
That's a good point. An actionable change would be adding a bounds check ensuring `sp` is not greater than the start of the stack region.

With that change edge cases will at least result in a failure.